compiler/noirc_frontend/src/elaborator/enums.rs

```diff
@@ -20,13 +20,87 @@ use crate::{
         function::{FuncMeta, FunctionBody, HirFunction, Parameters},
         stmt::{HirLetStatement, HirPattern, HirStatement},
     },
-    node_interner::{DefinitionId, DefinitionKind, ExprId, FunctionModifiers, GlobalValue, TypeId},
+    node_interner::{
+        DefinitionId, DefinitionKind, ExprId, FunctionModifiers, GlobalValue, NodeInterner, TypeId,
+    },
     signed_field::SignedField,
     token::Attributes,
 };
 
 use super::Elaborator;
 
+struct MatchCompiler<'elab, 'ctx> {
+    elaborator: &'elab mut Elaborator<'ctx>,
+    has_missing_cases: bool,
+    // We iterate on this to issue errors later so it needs to be a BTreeMap (versus HashMap) to be
+    // deterministic.
+    unreachable_cases: BTreeMap<RowBody, Location>,
+}
+
+/// A Pattern is anything that can appear before the `=>` in a match rule.
+#[derive(Debug, Clone)]
+enum Pattern {
+    /// A pattern checking for a tag and possibly binding variables such as `Some(42)`
+    Constructor(Constructor, Vec<Pattern>),
+    /// An integer literal pattern such as `4`, `12345`, or `-56`
+    Int(SignedField),
+    /// A pattern binding a variable such as `a` or `_`
+    Binding(DefinitionId),
+
+    /// Multiple patterns combined with `|` where we should match this pattern if any
+    /// constituent pattern matches. e.g. `Some(3) | None` or `Some(1) | Some(2) | None`
+    #[allow(unused)]
+    Or(Vec<Pattern>),
+
+    /// An integer range pattern such as `1..20` which will match any integer n such that
+    /// 1 <= n < 20.
+    #[allow(unused)]
+    Range(SignedField, SignedField),
+
+    /// An error occurred while translating this pattern. This Pattern kind always translates
+    /// to a Fail branch in the decision tree, although the compiler is expected to halt
+    /// with errors before execution.
+    Error,
+}
+
+#[derive(Clone)]
+struct Column {
+    variable_to_match: DefinitionId,
+    pattern: Pattern,
+}
+
+impl Column {
+    fn new(variable_to_match: DefinitionId, pattern: Pattern) -> Self {
+        Column { variable_to_match, pattern }
+    }
+}
+
+#[derive(Clone)]
+pub(super) struct Row {
+    columns: Vec<Column>,
+    guard: Option<RowBody>,
+    body: RowBody,
+    original_body: RowBody,
+    location: Location,
+}
+
+type RowBody = ExprId;
+
+impl Row {
+    fn new(columns: Vec<Column>, guard: Option<RowBody>, body: RowBody, location: Location) -> Row {
+        Row { columns, guard, body, original_body: body, location }
+    }
+}
+
+impl Row {
+    fn remove_column(&mut self, variable: DefinitionId) -> Option<Column> {
+        self.columns
+            .iter()
+            .position(|c| c.variable_to_match == variable)
+            .map(|idx| self.columns.remove(idx))
+    }
+}
+
```

Comment:
All this was just moved from the bottom of the file to the top.

The only change is the addition of `original_body: RowBody` and `location: Location` in `Row`